### ISSUE TITLE
BG_MODE not changing properly

### ISSUE THREAD

**Issue by twVolc** [twVolc/PyCamPermanent#3]
BG_MODE seems to always be written as 5 in the final processing results. I think if the background processing is run once, it at least changes the BG_MODE during processing, but if Apply is clicked without running it also doesn't change BG_MODE.

**Comment by twVolc** [comment]
this is an old bug which may well have been fixed but I haven't closed - I wasn't using issues/github very well at the time. Worth checking if this still seems to be an issue

**Comment by twVolc** [comment]
I think this probably isn't the correct behavior, as usually clicking OK makes users think they have selected things? So I think on the OK click, even without clicking run the model should change and be set to use the current settings when processing data. Does this seem reasonable?

**Comment by ubdbra001** [comment]
Okay, thought not.
So, does the model need to be run in order for it to work?
i.e. should the 'OK' button automatically run the function that the 'Run' button uses, as well as closing the window. Or is it enough to change the parameter in the program, and then this is picked up downstream during processing?

**Comment by twVolc** [comment]
I think it should be enough to change it and it then be picked up downstream, as the background modelling is performed for every image pair during processing - so the single clicking of Run shouldn't be affecting how this works later (I hope). Possibly the only subtle reason for having it run on clicking OK is that that will ensure the optical depth image (in the Analysis Window) is accurately represented - i.e. I'm guessing if the modelling isn't Run when clicking OK then this image will remain representing the old  background model processing, so for Users it could be misleading. So perhaps clicking OK we do want it to Run - but we want it to Run and immediately then close the Background Model GUI frame, rather than having it remain and plotting the results there (as happens with 'Run'). 

I guess the optimal thing to do (so we aren't running and slowing things down when not necessary) is only have it run the model again if the Background Model has changed since it was last run (either through 'Run' or any other way), but this would require tracking the most recent model background value to decide on if we need to re-run it or not. Whether that's straightforward or is a bit more hassle than it's worth at this stage - i don't know.

**Comment by ubdbra001** [comment]
Okay, I'll have a go at doing that.

Do you think it would be worth adjusting the names of the buttons? Maybe change 'Run' to something like 'Preview'?

**Comment by twVolc** [comment]
That sounds like a good idea - I hadn't thought of that!

**Comment by ubdbra001** [comment]
It looks like the function being called when pressing 'OK' is the same as when closing the widow using the 'X' button (From line 2777 in `figures_analysis.py`:
```python
def close_window(self):
    """Restore current settings"""
    if pyplis_worker.bg_pycam:
        self.bg_mode = 7
    else:
        self.bg_mode = pyplis_worker.plume_bg_A.mode
    self.auto_param = pyplis_worker.auto_param_bg
    self.polyfit_2d_thresh = pyplis_worker.polyfit_2d_mask_thresh
    self.ref_check_lower = pyplis_worker.ref_check_lower
    self.ref_check_upper = pyplis_worker.ref_check_upper
    self.ref_check_mode = pyplis_worker.ref_check_mode
    self.frame.destroy()
    self.in_frame = False
```

So, I think creating a seperate function that does some of the above but also:
- Checks to see if the current selection for bg model is the same as the one already in the settings
  - If no, then runs the model (but maybe after the window has closed so the figure is only replotted for the analysis window)
  - If yes, leave it as is

The code that runs when you press 'Run' is (From line 2658 in `figures_analysis.py`):
```python
def run_process(self, reload_seq=True):
    """Main processing for background modelling and displaying the results"""
    self.gather_vars()
    pyplis_worker.model_background()
    self.frame.attributes('-topmost', 1)
    self.frame.attributes('-topmost', 0)
    if reload_seq:
        pyplis_worker.load_sequence(pyplis_worker.img_dir, plot=True, plot_bg=False)
    self.get_current_settings()
```

I'm not 100% sure which bits from the `run_process()` function are key (I suspect the bits referencing `pyplis_worker`), so if you have any pointers they'd be greatly appreciated.

**Comment by ubdbra001** [comment]
Think I got it, see #47

**Comment by twVolc** [comment]
Might be me using things incorrectly, but I'm currently having an issue getting different BG modes to work on the dev branch. It keeps printing out:
```
ERROR! When attempting pyplis background modelling: Cannot model tau image: mismatch with regard to dark correction state between plume image(s) and sky radiance image(s)
Reverting to basic rectangular background model. Note subsequent processing will attempt pyplis modelling again unless changed by the user.
```
so nothing changes. Will investigate more later when I get the chance - perhaps I've got something setup incorrectly, but I do have the dark image directory set.

**Comment by ubdbra001** [comment]
Hmmm, I did encounter this error on the `issue18` branch (and it should be fixed in the latest commit), but not the `dev` or the `bg_mode` branches.

**Comment by twVolc** [comment]
Ah, if you've caught this that's great to know!

**Comment by ubdbra001** [comment]
I think I have, but I'm not sure why it would show up in the `dev` version of pycam. Could you let me know the hash of the commit you were using?

**Comment by twVolc** [comment]
It should have been the most recent dev commit e6c5edc7c760bf9c7159d80820766472c9c990e7
There's definitely a chance I didn't have something setup correctly as I was in a bit of a rush, but thought I had it all loaded in properly.

**Comment by ubdbra001** [comment]
Huh, odd. Let me know if you encounter it again.
I encountered it when loading in a config. The dark image directory was set in the config, but hadn't been applied, so it wasn't updating the bg_A/B img. IT throws the error when it gets to the `get_tau_image` function in `model_background`.

**Comment by twVolc** [comment]
I think this was my data by the way - tried with a different dataset and it worked, so I need to investigate what my dataset is missing rather than anything being wrong with this code.

**Comment by twVolc** [comment]
I think this has all been fixed on dev by a combination of #47 and the larger config file changes #65. Will be fixed on master with #86
